### DISCUSSION TITLE
refresh Alpine V3 after every livewire request also needed for liveview?

### DISCUSSION THREAD

**Discussion by dvic** [livewire/livewire#3392] [4. General]
We're having problems with upgrading to V3 with Phoenix LiveView and I wonder if it has anything to do with the following change in livewire:

https://github.com/livewire/livewire/pull/2960/files#diff-7ce6ac695f53fd2e2c50f1f962b00fb2da5cdf2576a1c1d6eeacca29f34b6024R16-R27

Is this something specific to Livewire or do we also have to apply the same logic in Phoenix LiveView?

(this is related to #3279)

**Comment by joshhanley** [comment]
@dvic there were some extra changes we had to implement after that PR as there were a couple of issues with it.

What issues are you guys having? Happy to help dig and see if we can get it sorted 🙂

If you can share your V2 support code, and also what you currently have for V3 support, we can see if there are other changes needed to get it all running smoothly.

**Reply by dvic** [reply]
We are seeing an issue with popups not working anymore once we navigate to another page (in Safari the popup briefly flickers, in other browsers it doesn't show), we have markup similar to this:

```html
<div x-data="{ open: false }">
    <button type="button" @click="open = !open">click here</button>
    <div @click.away="open = !open"
         x-show="open"
         x-transition
         class="origin-top-right absolute right-0 mt-2 w-56 rounded-md shadow-lg bg-white
                ring-1 ring-black ring-opacity-5 focus:outline-none">
         CONTENTS HERE
    </div>
</div>
```

with the dom hooks setup like:

```js
let liveSocket = new LiveSocket("/live", Socket, {
  dom: {
    onBeforeElUpdated(from, to) {
      if (from._x_dataStack) {
        window.Alpine.clone(from, to);
      }
    },
  },
});
```

But I have come to believe this is not due to the dom hooks setup function. The issue is somehow related to the `x-transition` directive because once we remove that one, we don't see the problem anymore. It seems like the transition callbacks are applied twice or something like that. I tried to replicate the problem in a separate liveview project but was not successful thusfar, it has problem to do with a specific morphdom operation (that our project incurs) that causes this bug in Alpine.js.

I'm not sure how to proceed now, the best would be to have an example outside livewire and liveview but I'm not sure if I can trigger the particular problem in a JS-only project. In any case, I think I should open an issue in the alpine repo right?

**Reply by dvic** [reply]
For the curious ones, I have tried reproducing it here: https://jsbin.com/joramatifo/edit?html,js,console,output

**Reply by dvic** [reply]
I have managed to reproduce it in a fresh Phoenix LiveView project, I'll open an issue in the alpine and phoenix liveview repo (to figure out which morphdom operation runs here that causes alpine to break).

https://github.com/dvic/phoenix_issues/tree/alpinev3-issue

![image](https://user-images.githubusercontent.com/1214337/126635068-53b4bd26-ad14-4883-84ac-75cdb0220230.png)

**Comment by joshhanley** [comment]
@dvic just going back through old notifications and saw I missed this. Did you end up getting it sorted?

**Reply by dvic** [reply]
Hi! No, unfortunately we did not manage to figure this one out, we're still stuck on 2.x. Two days ago someone else also asked me on the liveview repo and came up with a workaround: https://github.com/phoenixframework/phoenix_live_view/issues/1546#issuecomment-913023487 

I did notice this: https://github.com/livewire/livewire/blob/5cffdbfe4204e0d932cde6f8c9bdf63f0928b4d4/js/component/SupportAlpine.js#L260

It seems like there is something special one must do when using x-show and transitions? But the weird thing is that this is for v2 if I read the code correctly, and with v2 we don't have any problems. I'm really curious to see if this is also happening with livewire but I did not have time to investigate this.

**Reply by joshhanley** [reply]
@dvic bugger!

Yeah that code is for V2 transitions. V3 support seems to just work haha.

I'm just going to try and replicate your issue in Livewire and see if it works or not.

Just to confirm, you are doing a redirect to a page that has that Alpine component with `x-transition` and `@click.outside` on it that when you click away, its not transitioning or hiding. Is that correct?

Can you also confirm for me, a `live_redirect` in LiveView, is that a normal full page refresh when it redirects, or is that an SPA style page redirect?

**Comment by dvic** [comment]
> I'm just going to try and replicate your issue in Livewire and see if it works or not.

Great! Thanks a lot!

> Just to confirm, you are doing a redirect to a page that has that Alpine component with x-transition and @click.outside on it that when you click away, its not transitioning or hiding. Is that correct?

The problem is slightly different: I can't even get the pop-up to open, it gets closed immediately (it blinks briefly). When I debugged the dom changes it seems that the x-transition is applied twice and in the wrong order, so that in the end `style="display: none"` is left, while the pop-up should be open. Here's a gif:

![recording](https://user-images.githubusercontent.com/1214337/132199003-4ce829f3-fc9a-48af-965a-5c7d09998e2f.gif)

> Can you also confirm for me, a live_redirect in LiveView, is that a normal full page refresh when it redirects, or is that an SPA style page redirect?

It's a SPA style page redirect, when I fully refresh the page, everything works normally.

**Reply by joshhanley** [reply]
@dvic ah right, I'm not sure I can mock that, as Livewire doesn't have built in SPA style redirect. We actually just make use of turbo/turbolinks if users require that.

Currently there are issues with that and Livewire with Alpine V3. Have a look at this PR #3151 which is a failing test for it.

Do you have any code the initialises Alpine after a redirect has happened? If so can you link to it?

Edit: Had linked wrong PR 😆

**Comment by dvic** [comment]
> Currently there are issues with that and Livewire with Alpine V3. Have a look at this PR #3151 which is a failing test for it.

Interesting! Will take a look there.

> Do you have any code the initialises Alpine after a redirect has happened? If so can you link to it?

No, I don't think this is possible currently with LiveView but I will have to check. This is all we currently do (when an element is updated, so this is probably also called when one SPA redirects):
https://github.com/dvic/phoenix_issues/blob/alpinev3-issue/assets/js/app.js#L47-L49

**Reply by dvic** [reply]
FYI: there are the events `phx:page-loading-start` and `phx:page-loading-stop`(used for showing the progress bar when live (SPA) navigation and form submits are triggered).

**Reply by joshhanley** [reply]
@dvic thanks for sharing.

Hmm ok, so with Alpine V3, it has a mutation observer running, which picks up on any dom changes and initialises components if necessary.

So I suspect that when a page change is happening, Alpine's not initialising correctly after the changes, or its actually initialising twice which is doubling up the handlers, which would explain why when you click it appears and disappears instantly. That suggest to me that Alpine is running (not erroring out), so it comes down to how Alpine is being initialised when doing page redirects.

Just to confirm, does morphdom manage changing the DOM when you do page redirects/changes?

One other test you could do, without doing a page redirect, is hide your Alpine component behind a LiveView conditional (if) and add another button to the screen which toggles the conditional to true, then test to see if the Alpine component works ok.

The intent is to see, if the Alpine component doesn't exist in the dom, then it gets added by morphdom, does it then function ok when not doing a page redirect.

If it does work ok, then investigation will need to happen around controlling Alpine when page loads happen. But if it fails, then it's more of a morphdom issue.

Hope this helps!

**Reply by dvic** [reply]
> Just to confirm, does morphdom manage changing the DOM when you do page redirects/changes?

Yes, the morphdom stuff is triggered in this case.

> One other test you could do, without doing a page redirect, is hide your Alpine component behind a LiveView conditional (if) and add another button to the screen which toggles the conditional to true, then test to see if the Alpine component works ok.
> 
> The intent is to see, if the Alpine component doesn't exist in the dom, then it gets added by morphdom, does it then function ok when not doing a page redirect.

If I do this, then it works without problems. I think the reason for this is that with this "toggle" button, the morphdom changes are "fine-grained", replacing the div/buttons etc. But with a redirect to another page, the whole wrapping div is replaced (just 1 morphdom change).

**Reply by joshhanley** [reply]
@dvic ah interesting! Good to know.

Yeah I suspect it has something to do with Alpine's mutation observer and the whole page being changed causing the issues.
